executor/analyze.go

```diff
@@ -502,16 +518,454 @@ type AnalyzeFastExec struct {
+	}
+}
+
+func (e *AnalyzeFastExec) buildHist(ID int64, collector *statistics.SampleCollector, tp *types.FieldType) (*statistics.Histogram, error) {
+	// build collector properties.
+	collector.UpdateTotalSize()
+	collector.Samples = collector.Samples[:e.sampCursor]
+	collector.Count = int64(e.sampCursor)
+	for _, sample := range collector.Samples {
+		if sample.Value.IsNull() {
+			collector.NullCount++
+		} else {
+			err := collector.FMSketch.InsertValue(e.ctx.GetSessionVars().StmtCtx, sample.Value)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+	}
+	// TODO: build collector's CMSketch here.
+
+	// build histogram.
+	hist, err := statistics.BuildColumnWithSamples(e.ctx, int64(e.maxNumBuckets), ID, collector, tp, int64(e.rowCount))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return hist, nil
+}
+
+func (e *AnalyzeFastExec) runTasks() ([]*statistics.Histogram, []*statistics.CMSketch, error) {
+	errs := make([]error, e.concurrency)
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	// collect column samples and primary key samples and index samples.
+	length := len(e.colsInfo) + hasPKInfo + len(e.idxsInfo)
+	collectors := make([]*statistics.SampleCollector, length)
+	for i := range collectors {
+		collectors[i] = &statistics.SampleCollector{
+			FMSketch:      statistics.NewFMSketch(maxSketchSize),
+			MaxSampleSize: int64(MaxSampleSize),
+			CMSketch:      statistics.NewCMSketch(defaultCMSketchDepth, defaultCMSketchWidth),
+			Samples:       make([]*statistics.SampleItem, MaxSampleSize),
+		}
+	}
+
+	e.wg.Add(e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		bo := tikv.NewBackoffer(context.Background(), 500)
+		go e.handleSampTasks(bo, i, &errs[i], collectors)
+	}
+	e.wg.Wait()
+	for _, err := range errs {
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+	}
+
+	var err error
+	bo := tikv.NewBackoffer(context.Background(), 500)
+	err = e.handleScanTasks(bo, collectors)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	hists, cms := make([]*statistics.Histogram, length), make([]*statistics.CMSketch, length)
+	for i := 0; i < length; i++ {
+		if i < hasPKInfo {
+			hists[i], err = e.buildHist(e.pkInfo.ID, collectors[i], &e.pkInfo.FieldType)
+		} else if i < hasPKInfo+len(e.colsInfo) {
+			hists[i], err = e.buildHist(e.colsInfo[i-hasPKInfo].ID, collectors[i], &e.colsInfo[i-hasPKInfo].FieldType)
+		} else {
+			hists[i], err = e.buildHist(e.idxsInfo[i-hasPKInfo-len(e.colsInfo)].ID, collectors[i], types.NewFieldType(mysql.TypeBlob))
+		}
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		cms[i] = collectors[i].CMSketch
+	}
+	return hists, cms, nil
+}
+
 func (e *AnalyzeFastExec) buildStats() (hists []*statistics.Histogram, cms []*statistics.CMSketch, err error) {
-	// TODO: do fast analyze.
-	return nil, nil, nil
+	// To set rand seed, it's for unit test.
+	rander := rand.New(rand.NewSource(RandSeed))
+
+	// Only four rebuilds for sample task are allowed.
```

Comment:
1. The loop says it is five?
2. This is used handle region errors?

Reply:
Answer:
1. The number of rebuilds is four, is the five builds.
2. Yes.
3. Now keeping a constant is ok. If there are any other requests, we can change it to session variable or config item.